hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/UserMetricsUtil.java

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.metrics;
+
+import com.hazelcast.jet.aggregate.AggregateOperation;
+import com.hazelcast.jet.aggregate.AggregateOperation1;
+import com.hazelcast.jet.aggregate.AggregateOperation2;
+import com.hazelcast.jet.core.metrics.MetricsContext;
+import com.hazelcast.jet.core.metrics.ProvidesMetrics;
+import com.hazelcast.jet.function.BiConsumerEx;
+import com.hazelcast.jet.function.BiFunctionEx;
+import com.hazelcast.jet.function.BiPredicateEx;
+import com.hazelcast.jet.function.FunctionEx;
+import com.hazelcast.jet.function.SupplierEx;
+import com.hazelcast.jet.function.TriFunction;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+public final class UserMetricsUtil {
+
+    private static final ProvidesMetrics DUMMY_METRICS_PROVIDER = context -> { };
+
+    private UserMetricsUtil() {
+    }
+
+    public static ProvidesMetrics cast(Object metricsProviderCandidate) {
+        if (metricsProviderCandidate instanceof ProvidesMetrics) {
+            return (ProvidesMetrics) metricsProviderCandidate;
+        } else {
+            return DUMMY_METRICS_PROVIDER;
+        }
+    }
+
+    public static <T, U> BiPredicateEx<T, U> wrap(BiPredicateEx<T, U> biPredicateEx, Object metricsProviderCandidate) {
+        if (metricsProviderCandidate instanceof ProvidesMetrics) {
+            return new WrappedBiPredicateEx<>(biPredicateEx,
+                    Collections.singletonList((ProvidesMetrics) metricsProviderCandidate));
+        } else {
+            return biPredicateEx;
+        }
+    }
+
+    public static <T, R> FunctionEx<T, R> wrap(FunctionEx<T, R> functionEx, Object metricsProviderCandidate) {
```

Comment:
the candidate is only type Object, it can be constrained stronger

Reply:
I'm not sure it can... This particular wrap for example has usages with both `FunctionEx` and `PredicateEx` candidates. 
`extends ProvidesMetrics` is also not a good constraint, because most of the time candidates don't extend it and no wrapping will take place. 
If I'm misunderstanding something pls. provide an example of what you meant.